[Feature] Support vote fusion

THU17cyz opened the pull request:
Added the module which fuses 2d and 3d and generates 3 cues:

-  Semantic
-  Texture
- Geometric

THU17cyz commented:
Just updated as requested:

- Coords transformation during fusion as a separated module called Coord3DTransformation
- Replaced original code in vote fusion and point fusion with the module above
- Added unit test for this module
- Also updated unit test for Coord3DMode convert_point method, including a bug fix, and added TODO in that method because LIDAR implementation is still not correct
- Modified some docstring and comments

ZwwWayne commented on mmdet3d/models/fusion_layers/coord_transform.py:
```diff
@@ -0,0 +1,144 @@
+                 coords_type,
+                 img_meta=None,
+                 pcd_rotate_mat=None,
+                 pcd_scale_factor=None,
```
Should use the default value for this.
`pcd_scale_factor=1.0, pcd_trans_factor=0, pcd_horizontal_flip=False, pcd_vertical_flip=False`.

ZwwWayne commented:
Unit test failed.

ZwwWayne commented on mmdet3d/core/bbox/structures/coord_3d_mode.py:
```diff
@@ -245,7 +253,7 @@ def convert_point(point, src, dst, rt_mat=None):
         else:
             xyz = arr[:, :3] @ rt_mat.t()
 
-        remains = arr[..., 3:]
```
Is this change necessary?

THU17cyz replied:
This is not in accordance with the others, so I unified it.

THU17cyz replied:
if we want to support any shape of input pc, this function needs to be refactored

ZwwWayne commented on mmdet3d/models/fusion_layers/vote_fusion.py:
```diff
@@ -0,0 +1,251 @@
+                if 'img_crop_offset' in img_meta.keys() else 0)
+
+            # first reverse the data transformations
+            xyz_depth = coords_trans(seeds_3d_depth, 'HTSR', reverse=True)
```
This is a little bit hard code. If necessary, the img_meta should contain the information of 'HTSR'

Tai-Wang commented on mmdet3d/models/fusion_layers/vote_fusion.py:
```diff
@@ -0,0 +1,248 @@
+            bbox_num = bbox_2d_rescaled.shape[0]
+            seed_num = seed_3d_depth.shape[0]
+
+            # xyz_depth = seed_3d_depth.clone()
```
remove this unused code?

Tai-Wang commented on mmdet3d/models/fusion_layers/vote_fusion.py:
```diff
@@ -0,0 +1,248 @@
+            img_flatten += torch.tensor(
+                self.img_norm_cfg['mean'], device=img.device).unsqueeze(-1)
+            img_flatten = img_flatten - 128.0
+            img_flatten = img_flatten / 255.0
```
The normalization seems a little hard? If it is a must, at least we can merge these two lines.

THU17cyz replied:
Merged the last two lines. This is to keep the same as the official implementation

Tai-Wang commented on mmdet3d/models/fusion_layers/coord_transform.py:
```diff
@@ -0,0 +1,87 @@
+        'R': rotate_func,
+        'H': horizontal_flip_func,
+        'V': vertical_flip_func
+    }
```
The pipeline mapping is still like hard code, maybe at least we need to do some assertion/make it more compatible with any input (e.g. if op in pipeline does not belong to current supported pipeline_mapping?)

THU17cyz replied:
I added an assertion currently since making it compatible with any input seems not very necessary and last time Wenwei suggested not to over-engineer. Maybe you can take a look? @ZwwWayne

Tai-Wang reviewed (commented):
`transformation_3d_pipeline` is still not much elegant. Is there any other workaround? If not, we can make it more compatible and safe in the related implementation.

THU17cyz commented:
Updated according to the review.

Tai-Wang commented on mmdet3d/core/bbox/structures/utils.py:
```diff
@@ -122,7 +122,17 @@ def points_cam2img(points_3d, proj_mat):
     points_shape = np.concatenate([points_num, [1]], axis=0).tolist()
+    d1, d2 = proj_mat.shape[:2]
+    assert (d1 == 3 and d2 == 3) or (d1 == 3 and d2 == 4) or (d1 == 4
+                                                              and d2 == 4)
```
Please add instruction to the assertion like assert ..., "The shape of projection matrix is not supported."

Tai-Wang commented on mmdet3d/models/fusion_layers/coord_transform.py:
```diff
@@ -0,0 +1,88 @@
+        'V': vertical_flip_func
+    }
+    for op in list(pipeline):
+        assert op in pipeline_mapping.keys()
```
Please also add assertion information.

ZwwWayne commented on mmdet3d/models/fusion_layers/vote_fusion.py:
```diff
@@ -0,0 +1,249 @@
+
+            # normalization
+            # in accordance with the official implementation
+            img_flatten += torch.tensor(
```
The image should already been normalized after loading?

THU17cyz replied:
This is only to be in accordance with the official implementation, which subtracts 128 from the original image and divides it by 255.

THU17cyz replied:
The normalization used in 2d detection should make no difference. Shall I use this instead?

ZwwWayne commented on mmdet3d/models/fusion_layers/vote_fusion.py:
```diff
@@ -0,0 +1,249 @@
+            img_flatten = (img_flatten - 128.0) / 255.0
+
+            # take the normalized pixel value as texture cue
+            uv_flatten = uv_rescaled[:, 1].round().long() * \
```
Do they need to be long or int tensor after round() or floor() operation?

THU17cyz replied:
It must be long or the following op will fail

ZwwWayne commented on mmdet3d/models/fusion_layers/vote_fusion.py:
```diff
@@ -0,0 +1,249 @@
+
+            # normalization
+            # in accordance with the official implementation
+            img_flatten += torch.tensor(
```
may use image.new_tensor()

ZwwWayne commented on mmdet3d/models/fusion_layers/vote_fusion.py:
```diff
@@ -0,0 +1,249 @@
+
+                # convert from camera coords to depth coords
+                imvote = Coord3DMode.convert_point(
+                    imvote.view((-1, 3)).double(),
```
double or float?

ZwwWayne commented on tests/test_utils/test_coord_3d_mode.py:
```diff
@@ -62,21 +62,21 @@ def test_points_conversion():
 
     convert_depth_points = cam_points.convert_to(Coord3DMode.DEPTH)
     expected_tensor = torch.tensor([[
-        -5.2422e+00, -2.9757e-01, 4.0021e+01, 6.6660e-01, 1.9560e-01,
```
Ensure the correctness as this changes the expected_tensor.

THU17cyz commented:
@Tai-Wang @ZwwWayne I have just fixed all the issues mentioned before. Additionally, I added a unittest for vote fusion to make sure the rewrote VoteFusion is correct.

ZwwWayne commented on mmdet3d/core/bbox/structures/utils.py:
```diff
@@ -122,7 +122,20 @@ def points_cam2img(points_3d, proj_mat):
+
     points_shape = np.concatenate([points_num, [1]], axis=0).tolist()
+    assert len(proj_mat.shape) == 2, 'The dimension of the projection'\
+        'matrix should be 2 instead of %d.' % len(proj_mat.shape)
```
Use f-string

ZwwWayne commented on mmdet3d/core/bbox/structures/utils.py:
```diff
@@ -122,7 +122,20 @@ def points_cam2img(points_3d, proj_mat):
+    d1, d2 = proj_mat.shape[:2]
+    assert (d1 == 3 and d2 == 3) or (d1 == 3 and d2 == 4) or (
+        d1 == 4 and d2 == 4), 'The shape of the projection matrix'\
+        ' (%d*%d) is not supported.' % (d1, d2)
```
Use f-string

ZwwWayne commented on mmdet3d/datasets/pipelines/transforms_3d.py:
```diff
@@ -96,10 +96,15 @@ def __call__(self, input_dict):
                 ) < self.flip_ratio_bev_vertical else False
                 input_dict['pcd_vertical_flip'] = flip_vertical
 
+        if 'transformation_3d_pipeline' not in input_dict.keys():
```
May use `transformation_3d_flow`

ZwwWayne commented on mmdet3d/datasets/pipelines/transforms_3d.py:
```diff
@@ -96,10 +96,15 @@ def __call__(self, input_dict):
+
         if input_dict['pcd_horizontal_flip']:
             self.random_flip_data_3d(input_dict, 'horizontal')
+            input_dict['transformation_3d_pipeline'] += 'H'
```
Would list will be better than string?

ZwwWayne commented on mmdet3d/models/fusion_layers/coord_transform.py:
```diff
@@ -0,0 +1,89 @@
+    }
+    for op in list(pipeline):
+        assert op in pipeline_mapping.keys(), 'This data '\
+            'transformation op (%s) is not supported' % op
```
Use f-string

ZwwWayne commented on mmdet3d/models/fusion_layers/vote_fusion.py:
```diff
@@ -0,0 +1,240 @@
+                if 'scale_factor' in img_meta.keys() else [1.0, 1.0])
+
+            img_flip = img_meta['flip'] if 'flip' in \
+                img_meta.keys() else False
```
we can simply use `img_meta` rather than `img_meta.keys()`

ZwwWayne commented on mmdet3d/models/fusion_layers/vote_fusion.py:
```diff
@@ -0,0 +1,240 @@
+
+            img_scale_factor = (
+                seed_3d_depth.new_tensor(img_meta['scale_factor'][:2])
+                if 'scale_factor' in img_meta.keys() else [1.0, 1.0])
```
img_meta.keys() -> img_meta

ZwwWayne commented on mmdet3d/models/fusion_layers/vote_fusion.py:
```diff
@@ -0,0 +1,240 @@
+                # ray angle
+                ray_angle = seed_3d_expanded + imvote
+                ray_angle /= torch.sqrt(torch.sum(ray_angle**2, -1) +
+                                        1e-6).unsqueeze(-1)
```
May create a variable named EPS = 1e-6, and use EPS in these places.

ZwwWayne commented on mmdet3d/models/fusion_layers/coord_transform.py:
```diff
@@ -0,0 +1,89 @@
+        translate_func = partial(pcd.translate, trans_vector=pcd_trans_factor)
+        rotate_func = partial(pcd.rotate, rotation=pcd_rotate_mat)
+
+    # "T" stands for translation;
```
This note should be put in the Note section in the docstring
